### DISCUSSION TITLE
Jupynium x Nixos

### DISCUSSION THREAD

**Discussion by Gilk260** [kiyoon/jupynium.nvim#101] [Q&A] [answered]
Hello there.

I'm currently using `NixOs` and I have some trouble to make Jupynium working with it.
Just to resume what is NixOs, it is a read only file-systems, so each packages need to be packaged in order to be downloaded and installed.
Here are the different problems I have encountered:
- `pip` doesn't work with NixOs, need to be packaged;
- I packaged the library, but if I try to `:JupyniumStartAndAttachToServerInTerminal` in neovim and it failed:
`/nix/store/b9mkn6wrfrkkyaiix1p4rbipamvmyi1f-python3-3.11.6-env/bin/python3.11: No module named jupynium
[Process exited 1]
`
But when I run an interactive python session, I can import the module with no problem. Maybe neovim is starting with an other instance of Python3 but I don't really understand how it works and if it is really neovim which is starting Python3 or the command used.
- I also try to start by hand jupynium in a shell:
```sh
$ jupynium

Traceback (most recent call last):
  File "/nix/store/650k6dabzjmsgzw447y8y7xnbgz0zwjc-python3.11-jupynium-0.2.1/bin/.jupynium-wrapped", line 9, in <module>
    sys.exit(main())
             ^^^^^^
  File "/nix/store/9la51nnb4fnnx6cvmw2v7fznqgqqsnpq-python3-3.11.6-env/lib/python3.11/site-packages/jupynium/cmds/jupynium.py", line 466, in main
    q = persistqueue.UniqueQ(persist_queue_path)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/nix/store/9la51nnb4fnnx6cvmw2v7fznqgqqsnpq-python3-3.11.6-env/lib/python3.11/site-packages/persistqueue/sqlbase.py", line 405, in __init__
    self._init()
  File "/nix/store/9la51nnb4fnnx6cvmw2v7fznqgqqsnpq-python3-3.11.6-env/lib/python3.11/site-packages/persistqueue/sqlqueue.py", line 60, in _init
    super(SQLiteQueue, self)._init()
  File "/nix/store/9la51nnb4fnnx6cvmw2v7fznqgqqsnpq-python3-3.11.6-env/lib/python3.11/site-packages/persistqueue/sqlbase.py", line 413, in _init
    os.makedirs(self.path)
  File "<frozen os>", line 225, in makedirs
OSError: [Errno 30] Read-only file system: '/nix/store/9la51nnb4fnnx6cvmw2v7fznqgqqsnpq-python3-3.11.6-env/lib/python3.11/site-packages/jupynium/jupynium_persist_queue'
Exception ignored in: <function SQLiteBase.__del__ at 0x7f104cac4720>
Traceback (most recent call last):
  File "/nix/store/9la51nnb4fnnx6cvmw2v7fznqgqqsnpq-python3-3.11.6-env/lib/python3.11/site-packages/persistqueue/sqlbase.py", line 453, in __del__
    self._getter.close()
    ^^^^^^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'close'
```
I understand that you are trying to create or remove a folder, but with NixOs this is not possible.
- Alternative virtual env with `venv` or `conda`: Not a good idea, a lot of problem specific of NixOs, I already tried but not something we want to go.

Here is my `init.lua` file.
``` lua
return {
  -- other plugins
  {
    "kiyoon/jupynium.nvim",
    -- build = "pip3 install --user .",
    -- build = "conda run --no-capture-output -n jupynium pip install .",
    -- enabled = vim.fn.isdirectory(vim.fn.expand "~/miniconda3/envs/jupynium"),
  },
}
```

I have commented all the building part, like I said `pip` with NixOs is not working.

If you have any suggestions or you a solution, I'm all ears out.

**Comment by kiyoon** [comment]
Hi, in Jupynium config you can give the path to python binary location. Can you try this?

Also, in Jupynium there is a persistent queue that saves the last running Jupynium pid to a file. This is because it makes it easier to make jupynium not duplicate its run and instead it attaches to the originally running process thus you have one firefox open with multiple tabs.

I used a persistent queue with file rather than memory because it is more OS-agnostic. At least it works on Windows, Mac and Linux. I didn't know it wouldn't work on NixOS, which is unfortunate

**Reply by Gilk260** [reply]
`lua/jupynium/options.lua` with : `python_host = vim.g.python3_host_prog` ? Like I should hardcode the path to my python with the library installed ?

In my `settings.lua` file I have declared this :
```lua
vim.g.python3_host_prog = system("which python3")
``` It should already take the right python interpreter ?

**Reply by kiyoon** [reply]
I'm not entirely sure if jupynium's python path follows the setting you set, so it's good to ensure that the value is what you want. I assume that it is a virtual env / system python difference.

Does `which python3` give you `/nix/store/b9mkn6wrfrkkyaiix1p4rbipamvmyi1f-python3-3.11.6-env/bin/python3.11`? From the shell which you said it worked, is it the same python?

**Reply by Gilk260** [reply]
When I do `which python3` I got another nix path, like in `Nixos` likes to create multiple 'environments' in order to divide the system and ensure there are no conflicts or other things like that. I can't really explain why is there 2 Python, I think it is to ensure that the boot is not corrupted by some random library I would have installed, but the system-wide Python is the one with all libraries installed and it isn't the one launched with `:JupyniumStartAndAttachToServerInTerminal`. All other commands don´t do anything, I presume it's the same problem.
So where do I need to hardcode the path to the python I want to use ? Which file ?

**Reply by kiyoon** [reply]
You should config like this (you can find this in README.md)

```lua
require("jupynium").setup({
  --- For Conda environment named "jupynium",
  -- python_host = { "conda", "run", "--no-capture-output", "-n", "jupynium", "python" },
  python_host = vim.g.python3_host_prog or "python3",
})
```

**Reply by Gilk260** [reply]
I try with my plugins file like this (I'm using lazy vim):
```lua
return {
...
  {
    "kiyoon/jupynium.nvim",
    config = function()
      require("jupynium").setup({
        python_host = vim.g.python3_host_prog or "python3",
      })
    end,
    -- build = "pip3 install --user .",
    -- build = "conda run --no-capture-output -n jupynium pip install .",
    -- enabled = vim.fn.isdirectory(vim.fn.expand "~/miniconda3/envs/jupynium"),
  },
}
```
But still got the same error.

Here is a little screenshot.
![image](https://github.com/kiyoon/jupynium.nvim/assets/67020423/16b26c75-1cc9-4dd7-b63e-3aebd3a1587b)
I'm sorry I didn't mention this, in the screenshot we can see `nbclassic' '--jupyter_command' 'jupyter'` I think it might wanna start some sub-process with those commands and try to launch it with the wrong Python.

**Reply by Gilk260** [reply]
@kiyoon do you have any guess on this ?

**Reply by Gilk260** [reply]
Ok I have a new output, concerning the `persist_queue` I guess:
```sh
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "/nix/store/h13ryyffxxl3j429si145qdmqgj5jz83-python3-3.11.6-env/lib/python3.11/site-packages/jupynium/__main__.py", line 4, in <module>
    main()
  File "/nix/store/h13ryyffxxl3j429si145qdmqgj5jz83-python3-3.11.6-env/lib/python3.11/site-packages/jupynium/cmds/jupynium.py", line 466, in main
    q = persistqueue.UniqueQ(persist_queue_path)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/nix/store/h13ryyffxxl3j429si145qdmqgj5jz83-python3-3.11.6-env/lib/python3.11/site-packages/persistqueue/sqlbase.py", line 405, in __init__
    self._init()
  File "/nix/store/h13ryyffxxl3j429si145qdmqgj5jz83-python3-3.11.6-env/lib/python3.11/site-packages/persistqueue/sqlqueue.py", line 60, in _init
    super(SQLiteQueue, self)._init()
  File "/nix/store/h13ryyffxxl3j429si145qdmqgj5jz83-python3-3.11.6-env/lib/python3.11/site-packages/persistqueue/sqlbase.py", line 413, in _init
    os.makedirs(self.path)
  File "<frozen os>", line 225, in makedirs
OSError: [Errno 30] Read-only file system: '/nix/store/h13ryyffxxl3j429si145qdmqgj5jz83-python3-3.11.6-env/lib/python3.11/site-packages/jupynium/jupynium_persist_queue'
Exception ignored in: <function SQLiteBase.__del__ at 0x7f1a1fa9c9a0>
Traceback (most recent call last):
  File "/nix/store/h13ryyffxxl3j429si145qdmqgj5jz83-python3-3.11.6-env/lib/python3.11/site-packages/persistqueue/sqlbase.py", line 453, in __del__
    self._getter.close()
    ^^^^^^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'close'

[Process exited 1]
```
You said earlier `I used a persistent queue with file rather than memory because it is more OS-agnostic.` But like I said in NixOs, we can't write directly in the file systems. Is that possible to change the path where you write those files ?

**Reply by kiyoon** [reply]
Maybe we can use the temp directory. Can you try changing that and see if it works? I have no idea what path is writable and what is not

**Reply by Gilk260** [reply]
Can you tell me where i need to make those changes please?
And in usually with NixOs, it is better to write in HOME, because it's for user but some path are writtable for configuration but i won't explain hère, I may do a PR After aller this aventure XD

**Reply by kiyoon** [reply]
Sorry for the delay. You can modify the path here. The reason I put it in the package dir is that I wanted it to be cross-platform and each OS has a different directory system, thus I thought the installed path was the safest place. I didn't consider NixOS back then.

https://github.com/kiyoon/jupynium.nvim/blob/5595ed8ddf4cbdccf8ac139ead5e315cceeeedfc/src/jupynium/definitions.py#L6

**Reply by nitori** [reply]
You might want to check out [platformdirs](https://pypi.org/project/platformdirs/) for that, and then maybe use the user_data_dir (or user_cache_dir if it's only temporary data).

In a lot of linux like systems, python libraries can get installed into root-only locations (like /usr/lib or /usr/local/lib etc.). So modifying anything there as a regular user is not possible (though you should probably use a venv anyway).

Not sure how important that file is, but an upgrade of the library would likely nuke anything stored in the package dir anyway.

(Got here by someone asking for help in a python discord help-channel about this issue - I don't use this library myself)

Greetings
Nitori

**Reply by kiyoon** [reply]
@nitori seems like it needs a fix. Thanks for the suggestion! The cache dir seems good

**Reply by kiyoon** [reply]
Let me know what happens when you use #102 on NixOS. Thanks!

**Reply by Gilk260** [reply]
I'm using the Pypi package in order to have the module, and I don't know how to test this specific branch.
If @nitori knows how to test it on NixOs, it will be very graceful (and for my knowledge I would love to know how to test it also)

**Reply by nitori** [reply]
Something like this should probably work to install the PR:

`pip install -U git+https://github.com/kiyoon/jupynium.nvim.git@refs/pull/102/merge`

**Reply by Gilk260** [reply]
Unfortunately I can't do anything with pip on `NixOs`

**Reply by kiyoon** [reply]
I merged the PR, but I need to leave it for others to test it for at least 2 weeks before updating the version.

**Reply by Gilk260** [reply]
Hello @kiyoon, any news concerning the update of the pypi package ?

**Reply by kiyoon** [reply]
@Gilk260 Thanks for the reminder. Will update soon :)

**Reply by kiyoon** [reply, marked as answer]
@Gilk260 Done! Please check it out :) I also made a CI automated deployment so it should be easier to upload to PyPI for later versions.

**Reply by Gilk260** [reply]
It work **nicely** :heart:

![image](https://github.com/kiyoon/jupynium.nvim/assets/67020423/dc97ba1f-65b0-4088-b19d-2480fb1d1a6e)

Next step to package this thing for NixOs, but I think I will do it my self.

Thanks a lot, I love your package, like it's been nearly 1 year I'm seeing it and I couldn't use it

**Reply by kiyoon** [reply]
Oh my god! I'm sorry you had to wait that long :(( but thanks a lot for not forgetting it and trying it out!  
If you'd like to contribute on packaging for Nix, that'd be helpful for the community 👍 I don't know how it works but if you can even share a document about your procedure, that'd be good for the future reference.
